Sources/TuistKit/Commands/ScaffoldCommand.swift

```diff
@@ -0,0 +1,160 @@
+import TuistScaffold
+import TuistSupport
+
+enum ScaffoldCommandError: FatalError, Equatable {
+    var type: ErrorType { .abort }
+
+    case templateNotFound(String)
+    case templateNotProvided
+    case nonEmptyDirectory(AbsolutePath)
+
+    var description: String {
+        switch self {
+        case let .templateNotFound(template):
+            return "Could not find template \(template)"
+        case .templateNotProvided:
+            return "You must provide template name"
+        case let .nonEmptyDirectory(path):
+            return "Can't generate a template in the non-empty directory at path \(path.pathString)."
+        }
+    }
+}
+
+class ScaffoldCommand: NSObject, Command {
+    // MARK: - Attributes
+
+    static let command = "scaffold"
+    static let overview = "Generates new project based on template."
+    private let listArgument: OptionArgument<Bool>
+    private let pathArgument: OptionArgument<String>
+    private let templateArgument: PositionalArgument<String>
+    private var attributesArguments: [String: OptionArgument<String>] = [:]
+    private let subParser: ArgumentParser
+
+    private let templateLoader: TemplateLoading
+    private let templatesDirectoryLocator: TemplatesDirectoryLocating
+
+    // MARK: - Init
+
+    public required convenience init(parser: ArgumentParser) {
+        self.init(parser: parser,
+                  templateLoader: TemplateLoader(),
+                  templatesDirectoryLocator: TemplatesDirectoryLocator())
+    }
+
+    init(parser: ArgumentParser,
+         templateLoader: TemplateLoading,
+         templatesDirectoryLocator: TemplatesDirectoryLocating) {
+        subParser = parser.add(subparser: ScaffoldCommand.command, overview: ScaffoldCommand.overview)
+        listArgument = subParser.add(option: "--list",
+                                     shortName: "-l",
+                                     kind: Bool.self,
+                                     usage: "Lists available scaffold templates",
+                                     completion: nil)
+        templateArgument = subParser.add(positional: "template",
+                                         kind: String.self,
+                                         optional: true,
+                                         usage: "Name of template you want to use",
+                                         completion: nil)
+        pathArgument = subParser.add(option: "--path",
+                                     shortName: "-p",
+                                     kind: String.self,
+                                     usage: "The path to the folder where the template will be generated (Default: Current directory).",
+                                     completion: .filename)
+        self.templateLoader = templateLoader
+        self.templatesDirectoryLocator = templatesDirectoryLocator
+    }
+
+    func parse(with parser: ArgumentParser, arguments: [String]) throws -> ArgumentParser.Result {
```

Comment:
I find odd that we are extending the command protocol to add a `parse` method that only this `ScaffoldCommand` is leveraging. What about doing the following instead:
- We use the parser logic to parse the static arguments: list, path, template.
- In the run method we get all the arguments and strip the static ones. That will leave us with an array of arguments that we can validate.
- We can pass that list of arguments together with the already validated arguments to an utility that is responsible for doing a custom validation.

Something along the lines of:

```swift
func run(with arguments: ArgumentParser.Result) throws {
   // Get the static arguments
   let strippedArguments = ....
   if !shouldList {
      try scaffoldCommandArgumentsValidator.validate(arguments: strippedArguments, template: template)
   }
}
```

Reply:
In fact, since the validator will need to load the template to validate the arguments, the validate method can return the template:

```swift
let template = try validator.validate(arguments: strippedArguments, templateName: templateName)
```

Reply:
I see what you are saying, but the issue is that with this setup, running `parser.parse()` will throw an error because it does not know the arguments that the user is passing it to. That's why I need to pre-process it somehow, so not sure if I can postpone the validation to `run()`

Reply:
I am also leveraging this function in a WIP PR for signing to overcome the problem of registering subcommands to enable user running `tuist signing enrypt` [here](https://github.com/tuist/tuist/pull/1066/files#diff-e4d9c693eceb9561b7a724eb9e494940R34)

Reply:
> will throw an error because it does not know the arguments that the user is passing it to

Why do you think it'll fail?

Reply:
If the current `ArgumentParser` does not know the argument it is being passed, it will throw `unexpected argument`.
For example if I run now `tuist generate my_unregistered_argument`, it will fail with `Error: unexpected argument my_unregistered_argument; use --help to list available arguments`

Reply:
That's why I need to register the arguments from the template *before* running `parser.parse()`

Reply:
I think there are 2 types of arguments, static and dynamic:
- **Static:** template, path, list
- **Dynamic:** the arguments that will substitute the attributes in the template.

If we register the static ones and then call the command that you mentioned, `tuist scaffold template`, the parser should not error. At that point then we can add add our custom parsing while running the command. I'm not sure if I'm missing something but defining and parsing static arguments is something that we are doing in all the commands and it works.

Reply:
The static ones are not a problem, the problem is if I run
`tuist scaffold template --name MyFramework`. This will result in an error if we do not add some pre-process logic.